src/open-url.coffee

```diff
@@ -15,40 +17,103 @@ bookmarks = {}
 confirmations = [
 	'Done.'
 ]
+holdings = [
+	'Doing'
+]
 
 module.exports = (robot) ->
 	robot.http("#{firebaseUrl}/data/bookmarks.json?auth=#{firebaseAuth}")
 		.get() (err, res, body) ->
-			if err or res.statusCode isnt 200
+			if err? or res.statusCode isnt 200
 				msg.send 'Oops?'
 			else
 				bookmarks = JSON.parse body
 
-	open = (key, callback) ->
-		if bookmarks[key]?
-			robot.http(bookmarks[key]).get() (err, res, body) ->
-				callback (not err?) and res.statusCode is 200
-		else
-			callback false
+	###*
+	* This creates a function to shim the http response
+	* @param {function} post successes to
+	* @param {function} post failures to
+	* @return {function} Takes (err, res, body) and passes them to (resolve, reject)
+	###
+	createResolver = (resolve, reject) ->
+		(err, res, body) ->
+			if (not err?) and (res.statusCode is 200)
+				resolve(body)
+			else
+				reject(err ? new Error("StatusCode: #{res.statusCode}"))
 
-	bookmark = (key, value, callback) ->
-		bookmarks[key] = value
-		robot.http("#{firebaseUrl}/data/.json?auth=#{firebaseAuth}")
-			.patch(JSON.stringify({ bookmarks: bookmarks })) (err, res, body) ->
-				callback (not err?) and res.statusCode is 200
+	###*
+	* This creates a function to report errors
+	* @param {Object} Hubot msg object
+	* @return {function} Takes (Error) and reports it
+	###
+	createErrorReporter = (context) ->
+		(error) ->
+			robot.logger.error(error)
+			context.send('Something went wrong. Debug output logged')
 
-	robot.respond /open (\S+)$/i, (context) ->
-		open context.match[1], (done) ->
-			context.send if done then context.random confirmations else "Couldn't find " + context.match[1] + ' key.'
+	###*
+	* Attempt to visit the url referenced by key
+	* @param {string} url to visit
+	* @return {Promise} Handle output from the asynchronous
+	###
+	open = (url) ->
+		new Promise (resolve, reject) ->
+			robot.http(url).get() createResolver(resolve, reject)
+
+	###*
+	* Store a url in the cache and firebase
+	* @param {string} key under which to store it
+	* @param {string} bookmark to store
+	* @return {Promise} Handle output from the asynchronous section
+	###
+	bookmark = (key, value) ->
+		new Promise (resolve, reject) ->
+			bookmarks[key] = value
+			robot.http("#{firebaseUrl}/data/.json?auth=#{firebaseAuth}")
+				.patch(JSON.stringify({ bookmarks: bookmarks })) createResolver(resolve, reject)
+
+	###*
+	* Extract a value from the context provided
+	* @param {Object} Hubot msg object
+	* @return {string} Value from the stored bookmarks
+	###
+	getValueFromContext = (context) ->
+		if bookmarks[context.match[1]]?
```

Comment:
I think this should check the key, not the value. I.e. `if context.match[1]?`.

Currently this `if` means that if you ask for a named bookmark that doesn't exist, it checks the room bookmark instead. It should give you an error. Right now a typo in the bookmark name can give you a totally misleading result by silently hitting the wrong url.